### ISSUE TITLE
seqable->seqable becomes seqable->seq

### ISSUE THREAD

**Issue by yuhan0** [borkdude/clj-kondo#657]
**version**
clj-kondo v2019.11.23
**platform**
macOS 10.14, native
**editor**
Emacs+flycheck

**problem**
Looking at the source for the clojure.core types, I noticed an odd `:seqable-out` type that was being applied to the return values of the following functions:
- `reverse`
- `map`
- `remove`
- `filter` 
- `keep`

via `seqable->seqable`:
- `next`
- `rest`
- `nfirst`
- `nnext`
- `butlast`

I believe all of these functions return concrete lazy-seqs, and should be specced accordingly with `:seq`. 

**repro**

```clj
(ns repro)

(subvec (map inc (range)) 0 1)
```
**expected behavior**
```
Type error: expected vector but provided seq.
```
**actual behavior**
No errors reported because `:seqable-out` could be a vector according to the type checker

Note: I'm willing to submit a PR for this

**Comment by borkdude** [comment]
Changing this to `:seq` is not correct since seqable -> seqable functions only guarantee to operate on something seqable, then they call seq on that and return something which is seqable again. E.g. it would not be incorrect if such a function would return nil or a vector, however, those are implementation details that you should not rely on. See @puredanger's comments here: https://github.com/borkdude/speculative/issues/45

**Comment by yuhan0** [comment]
I still feel like these changes are justified in the context of clj-kondo whose goal is to be a useful linting tool, even if it means leveraging some current implementation details?
Perhaps not for `speculative` which I guess has to take a stricter approach to follow the correct spec.

Even if some functions like `cycle` return a specific class type that avoids forcing a seq, practically speaking it's used exactly like a seq 100% of the time and could never be treated as eg. a vector. 

I realise this is judgement-based and I might be missing some additional subtleties. Feel free to close if it doesn't align with clj-kondo's objectives, thanks!

**Comment by borkdude** [comment]
@yuhan0 You can try the change and post the output of script/diff here, let’s see how it behaves.

**Comment by yuhan0** [comment]
Here's the raw terminal output of `script/diff` after replacing all instances of :seqable-out -> :seq, looks like it didn't make any difference:

```
Linting and writing output to /tmp/clj-kondo-diff/branch.txt
Cloning into 'clj-kondo'...
remote: Enumerating objects: 123, done.
remote: Counting objects: 100% (123/123), done.
remote: Compressing objects: 100% (79/79), done.
remote: Total 5246 (delta 48), reused 67 (delta 24), pack-reused 5123
Receiving objects: 100% (5246/5246), 6.28 MiB | 3.75 MiB/s, done.
Resolving deltas: 100% (2884/2884), done.
Linting and writing output to /tmp/clj-kondo-diff/master.txt
1825c1825
< linting took 11089ms, errors: 394, warnings: 1430
---
> linting took 11112ms, errors: 394, warnings: 1430
```

**Comment by borkdude** [comment]
@yuhan0 Since there are no regressions and maybe only benefits to doing this, could you enumerate a couple of examples where this would be useful instead of the current type annotation? Preferably in the form of a PR + unit tests.

**Comment by yuhan0** [comment]
I think the only core functions which are restricted to work on lists or vectors are `peek`, `pop` and `subvec`. These functions throw an error when passed a seq and would be caught by a return value of :seq but not :seqable-out.

```clojure
(peek (cycle [1 2 3])
(subvec (filter even? (range))
        10 20)
```
However `subvec` is currently untyped and signatures of peek/pop are incorrect, hope it's alright if I fix those specs along with said PR :)

**Comment by borkdude** [comment]
@yuhan0 Go for it.

**Comment by yuhan0** [comment]
Hmm, under the `could-be-relations` it says that a `:seq` could be a `:list`, and since `peek` and `pop` accept lists they don't produce warnings when given seqs.

I removed the :seq key from that map and ran `script/diff`, and again there were no regressions.

From my understanding, `could-be-relations` is a map from interfaces -> concrete types, it makes sense to say a `seqable` or `sequential` could be a list but not that a concrete `seq` could be a list.

**Comment by borkdude** [comment]
The could-be-relations are there to prevent warnings when given a more abstract type that could have been a concrete type.

E.g. when a function is spec-ed to receive a concrete list, but all it got is a seq that we don't know for sure is a list, it should imo not give a warning.  But maybe there are no functions that depend on an argument being a concrete list anywhere?

**Comment by yuhan0** [comment]
> [...] but all it got is a seq that we don't know for sure is a list

This is the bit that doesn't make sense to me - a `seq` is itself a concrete type that is distinct from a `list`.
 If there's a value specced as a "not-sure-if-seq-or-list", then that should be a `seqable` instead of a `seq`, and the bug would be in whichever function that tagged the value as a `seq` when it really should be a more general `seqable` type. (really hope I'm not splitting hairs here)

**Comment by borkdude** [comment]
```
$ clj
Clojure 1.10.1
user=> (seq? (list 1 2 3))
true
user=> (seq? (cons 1 [2 3]))
true
user=> (seq? (lazy-seq 1 [2 3]))
true
```

**Comment by borkdude** [comment]
The keywords are modelled after the predicate names in clojure, without the question marks.

**Comment by yuhan0** [comment]
```clojure
user=> (seq? (list 1 2 3))
true
```
That's really surprising to me, perhaps I have to read up more on the finer points of Clojure's type system...

**Comment by borkdude** [comment]
Seq is a central abstraction in Clojure, not a concrete type. For the purpose of linting we might change the return type of seqable->seqable functions to seq, but I'm not sure if this is a good move. For peek and pop we might introduce a new abstraction, :stack`, take a look at speculative for this.

**Comment by puredanger** [comment]
Sequences are a "logical list" abstraction. Lists are concrete lists (and thus also directly support the logical list abstraction).

**Comment by puredanger** [comment]
> when a function is spec-ed to receive a concrete list, but all it got is a seq that we don't know for sure is a list, it should imo not give a warning. But maybe there are no functions that depend on an argument being a concrete list anywhere?

I can't think of any core functions that depend specifically on concrete list returns (except maybe the `list` constructor). Even `list*` says that it returns a sequence.

**Comment by yuhan0** [comment]
I see, my misunderstanding came from thinking `:seq` referred to the lazy-seqs that are returned from functions like `map` and `filter`, instead of the ISeq abstraction. 

It seems like there's no way within the existing type system for a linter to rule out `(peek x)` if x was produced from a seq-returning function, which was what I was attempting to do.
